src/GitHub.Exports.Reactive/Collections/TrackingCollection.cs

```diff
@@ -0,0 +1,472 @@
+using System.ComponentModel;
+using System.Reactive.Linq;
+using System.Reactive.Concurrency;
+using System.Reactive.Disposables;
+using GitHub.Helpers;
+using System.Reactive;
+using System.Linq;
+using ReactiveUI;
+
+namespace GitHub.Collections
+{
+    public class TrackingCollection<T> : ObservableCollection<T>, ITrackingCollection<T>
+        where T : ICopyable<T>
+    {
+        enum TheAction
+        {
+            None,
+            Move,
+            Add,
+            Insert
+        }
+
+        CompositeDisposable disposables = new CompositeDisposable();
+        IObservable<Unit> source;
+        Func<T, T, int> comparer;
+        Func<T, int, IList<T>, bool> filter;
+        IScheduler scheduler;
+        List<T> original;
+
+        public TrackingCollection(Func<T, T, int> comparer = null, Func<T, int, IList<T>, bool> filter = null, IScheduler scheduler = null)
+        {
+            this.scheduler = scheduler ?? RxApp.MainThreadScheduler;
+            this.comparer = comparer;
+            this.filter = filter;
+            if (filter != null)
+                original = new List<T>();
+        }
+
+        public TrackingCollection(IObservable<T> source, Func<T, T, int> comparer = null, Func<T, int, IList<T>, bool> filter = null, IScheduler scheduler = null)
+            : this(comparer, filter, scheduler)
+        {
+            Listen(source);
+        }
+
+        public ITrackingCollection<T> Listen(IObservable<T> obs)
+        {
+            source = obs
+                .ObserveOn(filter != null ? TaskPoolScheduler.Default : scheduler)
+                .Select(t =>
+                {
+                    var list = filter == null ? this : original as IList<T>;
+
+                    var idx = list.IndexOf(t);
+                    if (idx >= 0)
+                    {
+                        var old = list[idx];
+                        var comparison = 0;
+                        if (comparer != null)
+                            comparison = comparer(t, old);
+                        list[idx].CopyFrom(t);
+
+                        if (comparer == null || comparison == 0)
+                            return Tuple.Create(TheAction.None, idx, idx); // no sorting to be done, just replacing the element in-place
+                        else
+                        {
+                            var i = idx;
+                            if (comparison < 0) // replacing element has lower sorting order, find the correct spot towards the beginning
+                                for (var pos = i - 1; i > 0 && comparer(list[i], list[pos]) < 0; i--, pos--)
+                                    Swap(list, i, pos);
+                            else if (comparison > 0) // replacing element has higher sorting order, find the correct spot towards the end
+                                for (var pos = i + 1; i < list.Count - 1 && comparer(list[i], list[pos]) > 0; i++, pos++)
+                                    Swap(list, i, pos);
```

Comment:
This is an implementation of insertion sort - http://www.sorting-algorithms.com/insertion-sort
It's the fastest if the list is already almost sorted and spends minimal memory doing it.
In the case where there's no filter, we're sorting the underlying `Items` property of the collection, which triggers collection changed events for the UI, so we're a bit stuck with it. If we are using a filter, we need to figure out what happens to the object (moved from x to y, removed, inserted, appended, not changed) so we can tell the filter what happened and which items to process, so we end up having to control the sorting process ourselves.

We can optimize the `var idx = list.IndexOf(t);` on line 56 for the (likely) case that the observable source is coming in sorted with the same sorting conditions that we're using, and save some time there.